A query with `select *` and multiple joins fails in validation

The following test:
```java
        @Test
        public void test_multipleJoinsWithSelectStar() {
            createMapping("m1", int.class, int.class);
            createMapping("m2", int.class, int.class);
            createMapping("m3", int.class, int.class);

            assertRowsOrdered("select * " +
                    "from m1 " +
                    "join m2 on m1.__key=m2.__key " +
                    "join m3 on m2.__key=m3.__key", rows(1));
        }
```

fails with: 
```
java.lang.AssertionError: fieldList must not be null, type = UNKNOWN
	at org.apache.calcite.rel.type.RelDataTypeImpl.getFieldList(RelDataTypeImpl.java:164)
	at org.apache.calcite.sql.validate.SqlValidatorImpl$Permute.<init>(SqlValidatorImpl.java:7031)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.expandStar(SqlValidatorImpl.java:664)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.expandSelectItem(SqlValidatorImpl.java:426)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validateSelectList(SqlValidatorImpl.java:4409)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validateSelect(SqlValidatorImpl.java:3652)
	at org.apache.calcite.sql.validate.SelectNamespace.validateImpl(SelectNamespace.java:64)
	at org.apache.calcite.sql.validate.AbstractNamespace.validate(AbstractNamespace.java:89)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validateNamespace(SqlValidatorImpl.java:1100)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validateQuery(SqlValidatorImpl.java:1071)
	at com.hazelcast.jet.sql.impl.validate.HazelcastSqlValidator.validateQuery(HazelcastSqlValidator.java:182)
	at org.apache.calcite.sql.SqlSelect.validate(SqlSelect.java:247)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validateScopedExpression(SqlValidatorImpl.java:1046)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validate(SqlValidatorImpl.java:752)
	at com.hazelcast.jet.sql.impl.validate.HazelcastSqlValidator.validate(HazelcastSqlValidator.java:177)
	at com.hazelcast.jet.sql.impl.parse.QueryParser.parse0(QueryParser.java:80)
	at com.hazelcast.jet.sql.impl.parse.QueryParser.parse(QueryParser.java:57)
	at com.hazelcast.jet.sql.impl.OptimizerContext.parse(OptimizerContext.java:137)
	at com.hazelcast.jet.sql.impl.CalciteSqlOptimizer.prepare(CalciteSqlOptimizer.java:232)
	at com.hazelcast.sql.impl.SqlServiceImpl.prepare(SqlServiceImpl.java:261)
	at com.hazelcast.sql.impl.SqlServiceImpl.query0(SqlServiceImpl.java:241)
	at com.hazelcast.sql.impl.SqlServiceImpl.execute(SqlServiceImpl.java:198)
	at com.hazelcast.sql.impl.SqlServiceImpl.execute(SqlServiceImpl.java:162)
	at com.hazelcast.sql.impl.SqlServiceImpl.execute(SqlServiceImpl.java:158)
	at com.hazelcast.sql.impl.SqlServiceImpl.execute(SqlServiceImpl.java:154)
	at com.hazelcast.sql.SqlService.execute(SqlService.java:89)
	at com.hazelcast.jet.sql.SqlTestSupport.assertRowsOrdered(SqlTestSupport.java:275)
	at com.hazelcast.jet.sql.impl.connector.map.SqlJoinTest$SqlInnerJoinTest.test_multipleJoinsWithSelectStar(SqlJoinTest.java:436)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
```

--- a/hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/validate/types/HazelcastTypeFactory.java
+++ b/hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/validate/types/HazelcastTypeFactory.java
@@ -210,7 +210,7 @@ public final class HazelcastTypeFactory extends SqlTypeFactoryImpl {
      *
      * @param targetType the target type to find the widest instance of.
      * @param types      the list of types to inspect.
-     * @return the found widest integer type.
+     * @return the widest integer type found.
      */
     private static RelDataType leastRestrictive(RelDataType targetType, List<RelDataType> types) {
         SqlTypeName typeName = targetType.getSqlTypeName();
@@ -235,12 +235,5 @@ public final class HazelcastTypeFactory extends SqlTypeFactoryImpl {
         assert maxBitWidthType.getSqlTypeName() == typeName;
 
         return HazelcastIntegerType.create((HazelcastIntegerType) maxBitWidthType, targetType.isNullable());
-    }
-
-    @Override
-    public RelDataType createJoinType(RelDataType... types) {
-        // Calcite's implementation of createJoinType() returns RelCrossType.
-        // RelCrossType.getSqlTypeName() returns null. We assume in several places that it's non null...
-        return createUnknownType();
     }
 }

--- a/hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SqlJoinTest.java
+++ b/hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SqlJoinTest.java
@@ -428,6 +428,18 @@ public class SqlJoinTest {
         }
 
         @Test
+        public void test_multipleJoinsWithSelectStar() {
+            createMapping("m1", int.class, int.class);
+            createMapping("m2", int.class, int.class);
+            createMapping("m3", int.class, int.class);
+
+            assertRowsOrdered("select * " +
+                    "from m1 " +
+                    "join m2 on m1.__key=m2.__key " +
+                    "join m3 on m2.__key=m3.__key", rows(1));
+        }
+
+        @Test
         public void test_innerJoinPartOfTheCompositeKey() {
             String leftName = randomName();
             TestBatchSqlConnector.create(

